### PR TITLE
Merge upstream 5.7.21 into Percona Server

### PR DESCRIPTION
**Pull request by percona-ysorokin**

### CODE REVIEW

**Comment by percona-ysorokin** [comment]
Jenkins build link (--big-test):
https://jenkins.percona.com/view/5.7/job/mysql-5.7-param/1588/

The following 4 tests already re-recorded (no new Jenkins run)
`tokudb.perfschema.start_server_tokudb`
`tokudb.parts.partition_alter4_tokudb`
`rocksdb.type_json`
`rocksdb.index_merge_rocksdb2`

**Comment by percona-ysorokin** [comment]
@robgolebiowski 
Please, take a look at the `keyring_vault` part. These changes were necessary because of the extensions in the `keyring` API
WL #9769 "Keyring migration tool and new commercial keyring_encrypted_file plugin."
(commit mysql/mysql-server@4b4dd7a).
(see commit message for more details)

File: client/dump/sql_formatter.cc [hunk 1]
```diff
@@ -314,11 +314,18 @@ void Sql_formatter::format_dump_start(
     if (m_options->m_gtid_purged == GTID_PURGED_ON ||
         m_options->m_gtid_purged == GTID_PURGED_AUTO)
     {
-      if (!m_mysqldump_tool_options->m_dump_all_databases)
+      if (!m_mysqldump_tool_options->m_dump_all_databases &&
+          *((ulong*)&m_options->m_gtid_purged) == ((ulong)GTID_PURGED_AUTO))
```
Comment (laurynas-biveinis) [inline client/dump/sql_formatter.cc]: Shouldn't this excellent code be replaced with a straightforward comparison as above due to https://bugs.mysql.com/bug.php?id=89422 / https://github.com/percona/percona-server/commit/f105be761ccfe313c7c7526171dc716ab33122e7 ?

**Review by laurynas-biveinis — CHANGES_REQUESTED** [review]

**Comment by laurynas-biveinis** [comment]
Please include TDB-94 in the commit message. My review comment change should require only --suite=main --do-test=mysqlpump testing

File: client/dump/sql_formatter.cc (hunk 1 above)
Comment (percona-ysorokin) [reply client/dump/sql_formatter.cc]: Fixed

**Comment by percona-ysorokin** [comment]
New Jenkins run (--suite=main --do-test=mysqlpump --big-test)
https://jenkins.percona.com/view/5.7/job/mysql-5.7-param/1600/parameters/
Some nodes are marked red there despite the fact that all the tests passed successfully (could be another problem with parsing Jenkins results).

**Review by laurynas-biveinis — APPROVED** [review]

**Pull request merged**